0.2.0 Release (#12)
ConduitIO/conduit-site: guides/2022-04-05-build-a-conduit-connector.mdx

---
title: 'How to build a Conduit Connector'
slug: build-a-conduit-connector
author: '@anaptfox'
author_url: https://twitter.com/anaptfox
author_image_url: https://pbs.twimg.com/profile_images/1004707739405205504/BpvzOr9X_400x400.jpg
tag: ['Algolia Connector']
description: Learn how to build a Conduit Connector by implmenting an Algolia Destination Connector.
keywords: ['data', 'connector', 'algolia', 'conduit']
---

> # How to build a Conduit Algolia Connector

Reviewer: Shouldn't this say `How to build a Conduit Connector`?

> In this article we are going to walkthrough step by step how to build a Conduit connector. 

Reviewer: I'm definitely not the grammar authority around here so feel free to pick and choose my suggestions 😄 
```suggestion
In this article we are going to walk through how to build a Conduit connector step by step.
```

Conduit connectors are ... Conduit Connectors can be a source, destination, or both. 

For this example, we are going to build an Algolia destination connector. The goal of this connector is to give the user the ability to send data to Algolia. In the search engine world, this is called indexing. Since Conduit is a generic tool to move data between data infrastructure, with this new connector we can index data from any Conduit Source (PostgreSQL, Kafka, etc.).

> You may find this full example on [GitHub](/).

Reviewer: Missing link here. We need to make the Algolia example public first or remove the link if it's not ready by the time this needs to go out.

> Let's build.

Reviewer: Enthusiasm. 😐 
...
_Enthusiasm!_ 🤩 
```suggestion
Let's build!
```

<!--truncate-->

## How it works

In order to implement a Conduit connector. We need to write four methods:

1. Configure - Initialize and validate parameters
2. Open - Open connection to datastores, which is Algolia in this case.
3. Write - Write record to datastore.
4. Teardown - Clean up and close connection.

Learn more about the connector lifecycle in [The Guide to the Conduit Connector Lifecyle](/guides/guide-to-conduit-connector-lifecycle).

> Connectors can be created in many programming languages because Conduit Plugins leverage GRPC. For this example, we are going to use the Golang SDK. 

Reviewer: ```suggestion
Connectors can be created in many programming languages because Conduit Connectors leverage gRPC. For this example, we are going to use the Go [Connector SDK](https://github.com/conduitio/conduit-connector-sdk).
```

> Also, for this example, the datastore we are building the connector for is Algolia. Algola is TODO. As a destination, this connector allows for real-time syncing between any Conduit Source and Algolia.

Reviewer: Just pointing out the "TODO" here.

> To communicate with Algolia, we can use the [Algolia Golang library](https://www.algolia.com/doc/api-client/getting-started/install/go/?client=go).

Reviewer: As Haris already mentioned, the name of the language is "Go" 😉 
```suggestion
To communicate with Algolia, we can use the [Algolia Go library](https://www.algolia.com/doc/api-client/getting-started/install/go/?client=go).
```

### Connector File Structure

Next, let's take a look at an example connector file structure:

- `destination.go` - The main destination code.
- `object.go` - Helper functions to serialize records into JSON for Algolia
- `spec.go` - The specification defines connector metadata and parameters.

> - `cmd/algolia/main.go` - Entry point

Reviewer: ```suggestion
- `cmd/algolia/main.go` - Entrypoint
```

> - `cmd/algolia/main.go` - Entry point

Reviewer: I would add dots to the end of each bullet point. Or remove all dots, as long as it's consistent.

## Setup

Let's bootstrap our new connector. We are going to name our connector `conduit-connector-algolia`. It's best practice to use the `conduit-connector-NAME` format so the community can find connectors. 

1. Create a new folder for your connector:

```sh
mkdir -p /where/you/want/your/connector
```

2. Init Go Modules:

```sh
go mod init
```

3. Create the `main.go` entrypoint.

```
mkdir cmd/algolia && touch cmd/algolia/main.go
```
4. Create 

```
touch destination.go spec.go
```

5. Install Go dependencies

```

> go get

Reviewer: This won't do anything, I think what you mean is fetching the SDK and the Algolia library.
```suggestion
go get github.com/conduitio/conduit-connector-sdk
go get github.com/algolia/algoliasearch-client-go/v3
```

```

## Specification

First, let's build `spec.go`. The Specification

```go
package algolia

import sdk "github.com/conduitio/conduit-connector-sdk"

func Specification() sdk.Specification {
	return sdk.Specification{
		Name:        "algolia",
		Summary:     "A destination connector for Algolia",

> 		Description: "TBD",

Reviewer: Might want to add a longer description here.

		Version:     "v0.1.0",
		Author:      "Meroxa, Inc.",
		DestinationParams: map[string]sdk.Parameter{
			DestinationConfigAPIKey: {
				Required:    true,
				Description: "The API key for Algolia.",
			},
			DestinationConfigApplicationID: {
				Required:    true,
				Description: "The Application ID for Algolia.",
			},
			DestinationConfigIndexName: {
				Required:    true,
				Description: "The Algolia index where records get written into.",
			},
		},
		SourceParams: nil,
	}
}
```

> ## Entrypoint

Reviewer: I'd move this chapter after destination, otherwise we are referencing `algolia.NewDestination` which doesn't exist at this point.

Next, let's look at `cmd/algolia/main.go`. This is our entrypoint. 

```go
package main

import (
	algolia "github.com/conduitio/conduit-connector-algolia"
	sdk "github.com/conduitio/conduit-connector-sdk"
)

func main() {
	sdk.Serve(
		algolia.Specification,
		nil,
		algolia.NewDestination,
		)
}
```

> `sdk.Serve` is TODO. Notice `nil` as the second parameter. This is due to this connector being just a Destination.

Reviewer: "`sdk.Serve` starts the plugin and takes care of its whole lifecycle" - did you mean something like this?

## Destination

The `destination.go` implements the connector lifecycle methods. 

```go
...

type Destination struct {
	sdk.UnimplementedDestination

	config DestinationConfig
	index  *search.Index
}

type DestinationConfig struct {
	APIKey        string
	ApplicationID string
	IndexName     string
}
...
```

First we define two structs:

- Destination - This is the "storage" struct of your connector.
- DestinationConfig - This makes it easy to access connector config.

### Configure

> `Configure` is always the first function to be called in a connector and is used to store and validate the connector configuration. 

Reviewer: I would switch "store and validate", it should happen in the opposite order.
```suggestion
`Configure` is always the first function to be called in a connector and is used to validate and store the connector configuration. 
```

		return fmt.Errorf("%q is a required parameter", DestinationConfigAPIKey)
	}
	if destCfg.ApplicationID == "" {
		return fmt.Errorf("%q is a required parameter", DestinationConfigApplicationID)
	}
	if destCfg.IndexName == "" {
		return fmt.Errorf("%q is a required parameter", DestinationConfigIndexName)
	}

	d.config = destCfg
	return nil
}
```

### Open

`Open` is called after Configure to signal the plugin it can prepare to
start producing records. If needed, the plugin should open connections in
this function. 

```go

func (d *Destination) Open(ctx context.Context) error {
	client := search.NewClient(d.config.ApplicationID, d.config.APIKey)
	index := client.InitIndex(d.config.IndexName)
	d.index = index
	return nil
}
```

### Write

`Write` receives a Record and is supposed to write the record to the
destination right away.
```go

func (d *Destination) Write(ctx context.Context, record sdk.Record) error {
	 , err := d.index.SaveObject(Object(record))
	if err != nil {
		return fmt.Errorf("could not save object: %w", err)
	}

	sdk.Logger(ctx).Debug().
		Int("taskId", res.TaskID).
		Str("objectId", res.ObjectID).
		Msg("saved object")

	return nil
}

```

> The `Object` methods serializes the record. This is implemented in `object.go`.

Reviewer: `Object` is actually a type that changes the way the record is serialized to JSON. I think calling it a "method" is a bit misleading.

### Teardown

`Teardown` signals to the plugin that there will be no more calls to any
other function. 
```go

func (d *Destination) Teardown(ctx context.Context) error {
	// do nothing
	return nil
}
```

## Building

Now that we have all the needed methods implmented, we can build our new connector:

```
go build -o algolia cmd/algolia/main.go
```

> This will compile our connector to binary.

Reviewer: ```suggestion
This will compile our connector to a binary.
```

## Creating Pipeline with a new Connector

To use your new connector, perform the following steps:

1. Start Conduit.

> 2. Create Pipeline: 

Reviewer: ```suggestion
2. Create a Pipeline:
```

```curl

> curl 

Reviewer: Missing curl here.

```

> 3. Create Source

Reviewer: I would add a word or two describing that we are using a generator source that will produce random records every 5 seconds.

```json
{

> 	"type": "TYPE_SOURCE",
> 	"plugin": "builtin:generator",

Reviewer: Tabs vs spaces 😅 
```suggestion
    "type": "TYPE_SOURCE",
    "plugin": "builtin:generator",
```

    "config": {
        "readTime": "5s",
        "fields": "id:int",
    }
}
```

> 4. Create Destination

Reviewer: I'd also mention here that the `plugin` field can take either a path to the compiled connector binary or a name of a builtin plugin like `builtin:generator` (the plugin we are using for the source).

```json
{
	"type": "TYPE_DESTINATION",
	"plugin": "/path/to/algoila",
    "config": {
        "api_key": "",
        "application_id": "",
        "index_name": ""
    }
}
```

5. Start pipeline

After, starting the pipeline, you should see the following in the Algolia:

SCREENSHOT

## What's Next

> In this guide, we have covered the basics of how to build a connector for Conduit. The best way to learn more about what's possible with Conduit is to check out the [Conduit documentation](https://docs.conduit.io/), the [Conduit API documentation](https://docs.conduit.io/api/), and the [Conduit SDK](/).

Reviewer: https://docs.conduit.io/ still points to the main page instead of docs, maybe it would be better to change the link to https://www.conduit.io/docs/introduction/getting-started.

> In this guide, we have covered the basics of how to build a connector for Conduit. The best way to learn more about what's possible with Conduit is to check out the [Conduit documentation](https://docs.conduit.io/), the [Conduit API documentation](https://docs.conduit.io/api/), and the [Conduit SDK](/).

Reviewer: ```suggestion
In this guide, we have covered the basics of how to build a connector for Conduit. The best way to learn more about what's possible with Conduit is to check out the [Conduit documentation](https://docs.conduit.io/), the [Conduit API documentation](https://docs.conduit.io/api/), and the [Conduit Connector SDK](https://github.com/conduitio/conduit-connector-sdk).
```

You may also see examples of the methods above within the following connectors:

> - [PostgreSQL Connector](https://github.com/conduitio/conduit-connector-pg)

Reviewer: Broken link
```suggestion
- [PostgreSQL Connector](https://github.com/conduitio/conduit-connector-postgres)
```

---
title: 'How to build a Conduit Connector'
slug: build-a-conduit-connector
author: '@anaptfox'
author_url: https://twitter.com/anaptfox
author_image_url: https://pbs.twimg.com/profile_images/1004707739405205504/BpvzOr9X_400x400.jpg
tag: ['Conduit Algolia Connector']
description: Learn how to build a Conduit Connector by implmenting an Algolia Destination Connector.
keywords: ['data', 'connector', 'algolia', 'conduit']
---

# How to build a Conduit Algolia Connector

> In this article we are going to walkthrough step by step how to build a Conduit connector. Conduit connectors communicates with Conduit by either writing records into the pipeline (source connector) and/or the other way around (destination connector). 

Reviewer 2: Conduit connectors communicates - Conduit connectors communicate

	sdk "github.com/conduitio/conduit-connector-sdk"
)

func main() {
	sdk.Serve(
		algolia.Specification,
		nil,
		algolia.NewDestination,
		)
}
```

`sdk.Serve` starts the plugin and takes care of its whole lifecycle. Notice `nil` as the second parameter. This is due to this connector being just a Destination.

## Building

Now that we have all the needed methods implmented, we can build our new connector:

```
go build -o algolia cmd/algolia/main.go
```

This will compile our connector to a binary.

## Creating Pipeline with a new Connector

To use your new connector, perform the following steps:

1. [Start Conduit](https://github.com/ConduitIO/conduit#installation-guide).
2. Create Pipeline: 

```curl
curl -X 'POST' \
  'http://localhost:8080/v1/pipelines' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "config": {
    "name": "my-pipeline",
    "description": "This pipeline is for testing",
  }
}'
```
3. Create Source:

Since this is a Destination, we need to create a source. In this case, we can use the generator connector. It's a connector you can use for testing that will produce random records every 5 seconds.

> ```json

Reviewer 2: nitpick: json -> curl

curl -X 'POST' \
  'http://localhost:8080/v1/connectors' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
	"type": "TYPE_SOURCE",
	"plugin": "builtin:generator",
    "config": {
        "readTime": "5s",
        "fields": "id:int",
    }
}'
```

Note: The `plugin` field can take either a path to the compiled connector binary or a name of a builtin plugin like `builtin:generator`.

4. Create Algolia Destination:

```json
curl -X 'POST' \
  'http://localhost:8080/v1/connectors' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
	"type": "TYPE_DESTINATION",
	"plugin": "/path/to/algoila",
    "config": {
        "api_key": "",
        "application_id": "",
        "index_name": ""
    }
}'
```

5. Start pipeline

> After, starting the pipeline, you should see the following in the Algolia:

Reviewer 2: It appears we're missing the expected behavior in Algolia here.
